github-api-call-inventory/github-api-call-inventory.ipynb

```diff
@@ -0,0 +1,720 @@
+    "\n",
+    "The number of API calls drops from 345 to 65."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "id": "eba2722a",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "prs_per_hour_per_org = 76\n",
+       "total_is_project_private = 1\n",
+       "total_is_user_approved = 2\n",
+       "total_load_packit_yaml = 2\n",
+       "total_set_status = 60\n",
+       "grand_total = 65"
+      ]
+     },
+     "execution_count": 14,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "github_api_calls(\n",
+    "    chroots=6, \n",
+    "    rate_limit=5000, \n",
+    "    load_packit_yaml=2, \n",
+    "    builds_enabled=True, \n",
+    "    tests_enabled=True, \n",
+    "    load_packit_yaml_once=True, \n",
+    "    all_flags=False, \n",
+    "    is_private_once=True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "cea66f97",
+   "metadata": {},
+   "source": [
+    "The above is a 246% improvement even for the best case scenario, by dropping the number of API calls per pipeline from 120 to 35."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 15,
+   "id": "d4094d47",
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "prs_per_hour_per_org = 142\n",
+       "total_is_project_private = 1\n",
+       "total_is_user_approved = 2\n",
+       "total_load_packit_yaml = 2\n",
+       "total_set_status = 30\n",
+       "grand_total = 35"
+      ]
+     },
+     "execution_count": 15,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "github_api_calls(\n",
+    "    chroots=3,\n",
+    "    rate_limit=5000,\n",
+    "    load_packit_yaml=2,\n",
+    "    builds_enabled=True,\n",
+    "    tests_enabled=True,\n",
+    "    load_packit_yaml_once=True,\n",
+    "    all_flags=False,\n",
+    "    is_private_once=True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "7fed3b7e",
+   "metadata": {},
+   "source": [
+    "From this point on, the number of API calls could be further reduced only by having fewer status flags (merging flags for different chroots) or making fewer updates to these flags. Both of these could make the user experiance worse."
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "8745cba8",
+   "metadata": {},
+   "source": [
+    "Note, that the above assumes that pipelines are never running in parallel, which is probably not the case, for example, when PRs are updated while a previous pipeline is still in progress. In these situations old pipeline runs could still consume from the GitHub API rate limit (and maybe override statuses?). This could be solved by canceling Copr builds and test runs belonging to pipelines which become irrelevant ([packit/packit-service#5](https://github.com/packit/packit-service/issues/5))."
```

Comment:
Agree, this can be a big help, but a) not easy to do b) not always wanted (e.g. with branch commit push, I don't think it's clear if the user wants to skip the build triggered by a previous commit in case of it still being in progress).

Do you think it's worth investigating? Or, it's more beneficial to spend our time on other suggestions you've found?

Reply:
Why would it be difficult?

We have pipelines and from those we should know which builds/tests are still running. We could cancel those, by implementing something like `pipeline.cancel()`. 

Canceling previous pipelines could be optional, or not done at all in the case of branch pushes. In the case of PRs though it doesn't make sense to let old pipelines to finish.

Reply:
(Sorry for not getting back to this earlier..)

> Why would it be difficult?

I am not saying difficult, just that it requires a non-trivial change. (Compared to other enhancements that are more isolated.)

Reply:
So, not against it, but not sure if we should spend our time implementing that...;)